VERSIONING.md

```diff
@@ -0,0 +1,197 @@
+    by using the alternative. Otherwise, we can consider adding a flag (which is
+    now subject to a deprecation period of its own).
+* A release note at the introduction of the deprecation, along with the removal
+  of the behavior.
+* Any documentation that refers to the deprecated behavior in the Cosign
+  repository should have the same information as in the message.
+  * Other documentation will be updated on a best-effort basis.
+
+For a list of currently deprecated behavior, search the codebase.
+
+### Rationale/background
+
+Currently, many folks assume that the Cosign CLI follows semantic versioning. We
+have a bunch of breaking changes blocked on a 2.0 release. Even after that
+point, there are a number of proposed breaking changes/overhauls proposed. But
+there’s hesitancy to publish another major version. Further, semver doesn’t lend
+itself to a regular deprecation train (marking things deprecated, then ripping
+out after a fixed period of time); rather, in semver, you always have to wait
+for a new major version, which batches together a bunch of proposed deprecated
+behavior.
+
+Approaches considered include:
+
+* Follow semver. Bump major versions frequently (this is why Chrome is at
+  version 107 and Firefox is at 106, and gcloud is at version 409). I think this
+  is actually way more user-hostile; it effectively means "there are no
+  compatibility guarantees."
+* Follow semver. Avoid major version releases. I think that Sigstore and Cosign
+  are developing too rapidly to commit to avoid breaking things. Further, this
+  winds up preventing incremental improvements to the CLI in favor of huge
+  changes. For instance, right now with the v2 work underway the codebase is in
+  an awkward spot and would take a fair bit of work/branch management to push a
+  minor version release; realistically, the next release will be v2 even if we
+  have bug fixes we could push before then.
+
+## API support for old Cosign versions
+
+Starting with Cosign CLI v2, the Sigstore infrastructure should support old
+Cosign versions for one year past release, and six months past the availability
+of a suitable replacement. Any breakage here will be considered a bug and should
+be rolled back or fixed.
+
+We will run end-to-end tests for all prior supported minor versions.
+
+## Cosign API
+
+The Cosign API is versioned independently of the Cosign CLI. Version numbers are
+`MAJOR.MINOR.PATCH`. For now, there are **no stability guarantees**:
+
+* The library will undergo a serious refactor to make using it much
+  safer/easier.
+* Most callers (anybody not dealing with OCI registries/containers) will want
+  [sigstore-go][] instead.
+* At a future major version, the library will follow semantic versioning (and
+  this policy will be updated).
+* After that point, only code in pkg/ will be covered by semver; Cosign will
+  follow semver strictly.
+  * Most functionality will be in `internal/` or `cmd/` by default, and exposing it
```

Comment:
Question: @znewman01 Shouldn't `cmd` be focused on providing the CLI commands without exposing much functionality ? I was expecting to find here `internal/` and `pkg/`.

Reply:
That was  awkward wording. From the perspective of Go API consumers, the "public interface" is everything *except* `internal/` and `cmd/` (mostly `pkg/`). My point is that Go API users should not rely on `cmd/`.

I've reworded to:

>  Functionality should be private (either language-private or in a "private"
>  package like `internal/` or `cmd/`) by default, and exposing it publicly
>  will be a deliberate choice.

Hopefully that's clearer.